### PR TITLE
Switching mechanism for promotion policies

### PR DESCRIPTION
**Pull request by imjfckm**
Signed-off-by: Jan Musial <jan.musial@intel.com>

### CODE REVIEW

File: src/promotion/nhit/nhit_hash.c [hunk 1]
```diff
@@ -145,6 +145,7 @@ ocf_error_t hash_init(uint64_t hash_size, nhit_hash_t *ctx)
 	for (i_locks = 0; i_locks < new_ctx->hash_entries; i_locks++) {
 		if (env_rwsem_init(&new_ctx->hash_locks[i_locks])) {
 			result = -OCF_ERR_UNKNOWN;
+			i_locks--;
```
Comment (mmichal10) [inline src/promotion/nhit/nhit_hash.c]: i_locks is uint

Comment (imjfckm) [reply src/promotion/nhit/nhit_hash.c]: Fixed

File: src/promotion/nhit/nhit.c [hunk 2]
```diff
@@ -163,6 +178,11 @@ bool nhit_req_should_promote(ocf_promotion_policy_t policy,
 	bool result = true;
 	uint32_t i;
 	uint64_t core_line;
+	uint64_t occupied_cachelines =
+		ocf_metadata_get_cachelines_count(policy->owner) -
+		policy->owner->device->freelist_part->curr_size;
+	bool triggered = occupied_cachelines >
+		env_atomic64_read(&ctx->trigger_threshold);
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: Thought about that, but this way we won't 'touch' core lines in promotion policy before triggering. I think it is beneficial that as soon as the promotion policy is triggered we can start promoting instantly the cache lines that should be promoted. What do you think?

File: src/promotion/nhit/nhit.c [hunk 3]
```diff
@@ -41,6 +38,9 @@ ocf_error_t nhit_init(ocf_cache_t cache, ocf_promotion_policy_t policy)
 		goto dealloc_ctx;
 
 	env_atomic_set(&ctx->insertion_threshold, NHIT_THRESHOLD_DEFAULT);
+	env_atomic64_set(&ctx->trigger_threshold,
+			(NHIT_TRIGGER_DEFAULT *
+			 ocf_metadata_get_cachelines_count(cache)) / 100);
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: Done

File: src/promotion/nhit/nhit.c [hunk 4]
```diff
@@ -81,6 +81,21 @@ ocf_error_t nhit_set_param(ocf_promotion_policy_t policy, uint8_t param_id,
 		}
 		break;
 
+	case nhit_trigger_threshold:
+		if (param_value >= NHIT_MIN_TRIGGER &&
+				param_value < NHIT_MAX_TRIGGER) {
+			env_atomic64_set(&ctx->trigger_threshold,
+				(param_value *
+				ocf_metadata_get_cachelines_count(policy->owner)) / 100);
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: Done

File: src/promotion/nhit/nhit.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (imjfckm) [reply src/promotion/nhit/nhit.c]: True that, fixed

File: src/promotion/promotion.c [hunk 5]
```diff
@@ -84,25 +136,45 @@ ocf_error_t ocf_promotion_get_param(ocf_promotion_policy_t policy,
...
 
 bool ocf_promotion_req_should_promote(ocf_promotion_policy_t policy,
 		struct ocf_request *req)
 {
-	ocf_promotion_t type = policy->type;
+	ocf_promotion_t type;
 	bool result = true;
+	ocf_error_t lock_error;
+
+	lock_error = env_rwsem_down_read_trylock(&policy->type_lock);
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply src/promotion/promotion.c]: Apparently, as @michalwy says, we don't need those locks at all. It's meant to be used in IO path which has metadata locked. I only modified mgmt operations to take metadata lock as well and now it should all be fine.

**Pull request merged**